tsdb/index/postings.go

```diff
@@ -415,18 +415,44 @@ func (p *MemPostings) PostingsForLabelMatching(ctx context.Context, name string,
 		}
 	}
 
-	var its []Postings
-	count := 1
-	for _, v := range vals {
+	// We don't need to hold the mutex while we're matching,
+	// which can be slow (seconds) if the match function is a huge regex.
+	// Holding this lock prevents new series from being added (slows down the write path) and blocks the compaction process.
+	p.mtx.RUnlock()
+
+	for i, count := 0, 1; i < len(vals); count++ {
 		if count%checkContextEveryNIterations == 0 && ctx.Err() != nil {
-			p.mtx.RUnlock()
 			return ErrPostings(ctx.Err())
 		}
-		count++
-		if match(v) {
-			its = append(its, NewListPostings(e[v]))
+
+		if match(vals[i]) {
+			i++
+			continue
+		}
+
+		// Didn't match, bring the last value to this position, make the slice shorter and check again.
+		// The order of the slice doesn't matter as it comes from a map iteration.
+		vals[i], vals = vals[len(vals)-1], vals[:len(vals)-1]
+	}
+
+	// If none matched, no need to grab the lock again.
+	if len(vals) == 0 {
+		return EmptyPostings()
+	}
+
+	// Now `vals` only contains the values that matched, get their postings.
+	var its []Postings
+	p.mtx.RLock()
+	for _, v := range vals {
+		if refs, ok := e[v]; ok {
```

Comment:
I'm not sure it's safe accessing `e` after the previous lock was released (`p.m` could have changed in the meanwhile). We should re-read it from `p.m`.

To reduce such kind of risks, I would recommend moving the initial part of the function under a block, so that we keep the scope of variables limited, e.g.:

```
	var vals []string

	{
		p.mtx.RLock()

		e := p.m[name]
		if len(e) == 0 {
			p.mtx.RUnlock()
			return EmptyPostings()
		}

		// Benchmarking shows that first copying the values into a slice and then matching over that is
		// faster than matching over the map keys directly, at least on AMD64.
		vals = make([]string, 0, len(e))
		for v, srs := range e {
			if len(srs) > 0 {
				vals = append(vals, v)
			}
		}

		// We don't need to hold the mutex while we're matching,
		// which can be slow (seconds) if the match function is a huge regex.
		// Holding this lock prevents new series from being added (slows down the write path) and blocks the compaction process.
		p.mtx.RUnlock()
	}
```

Reply:
I see your concerns, @pracucci.

I'm _quite sure_ that there's no issue in reusing this that map reference as a map reference is just a pointer.

However, it's a valid concern that stuff from read-locked can escape into the next piece of code, and even though the code block would work, I opted for extracting the `labelValues(name string)` into a separate method. The extra method call shouldn't be a big deal here as we're going to call `match()` a few thousands of times later :)

code: https://github.com/prometheus/prometheus/pull/14286/commits/6582c9a731ae17aa816c34278ac1f8d8bb093f6d

Reply:
My understanding is that this can now be resolved because the code has been refactored and `e` is now set accesed just after the lock.

Reply:
Yes. Resolving it.